File: betsyro.java
// betsyro.java

// Betsyro is a software solution for online civil discourse and civic 
// engagement. It aims to empower those that are unable to be physically present
// at a political venue or otherwise participate in the democratic process.

// Copyright (C) 2015  Scott C. MacCallum
// scm@linux.com

// Copyright (C) 2015  Andrew Scott
// a.clayton.scott@gmail.com

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program.  If not, see http://www.gnu.org/licenses/.

import org.jibble.pircbot.*;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class betsyro extends PircBot {
    
    public betsyro() {
        this.setName("betsyro");
    }

    String[] badwords = new String[]{"shit", "poop", "ass","anal",
				     "anus",
				     "arse",
				     "ass",
				     "ballsack",
				     "balls",
				     "bastard",
				     "bitch",
				     "biatch",
				     "bloody",
				     "blowjob",
				     "blow job",
				     "bollock",
				     "bollok",
				     "boner",
				     "boob",
				     "bugger",
				     "bum",
				     "butt",
				     "buttplug",
				     "clitoris",
				     "cock",
				     "coon",
				     "crap",
				     "cunt",
				     "damn",
				     "dick",
				     "dildo",
				     "dyke",
				     "fag",
				     "feck",
				     "fellate",
				     "fellatio",
				     "felching",
				     "fuck",
				     "f u c k",
				     "fudgepacker",
				     "fudge packer",
				     "flange",
				     "Goddamn",
				     "God damn",
				     "hell",
				     "homo",
				     "jerk",
				     "jizz",
				     "knobend",
				     "knob end",
				     "labia",
				     "lmao",
				     "lmfao",
				     "muff",
				     "nigger",
				     "nigga",
				     "omg",
				     "penis",
				     "piss",
				     "poop",
				     "prick",
				     "pube",
				     "pussy",
				     "queer",
				     "scrotum",
				     "sex",
				     "shit",
				     "s hit",
				     "sh1t",
				     "slut",
				     "smegma",
				     "spunk",
				     "tit",
				     "tosser",
				     "turd",
				     "twat",
				     "vagina",
				     "wank",
				     "whore",
				     "wtf"};
    
    public void onMessage(String channel, String sender,
			  String login, String hostname, String message) {

        if (message.equalsIgnoreCase("What time is it?")) {
            String time = new java.util.Date().toString();
            sendMessage(channel, sender + ": The time is now " + time);
        }

	if (message.equalsIgnoreCase("Who is the mayor?")) {
            String time = new java.util.Date().toString();
	    sendMessage(channel, sender + ": Michael F. Brennan");
	}
  

	if (message.equalsIgnoreCase("What is the city managers email address?")) {
            String time = new java.util.Date().toString();
            sendMessage(channel, sender + ": citymanager@portlandmaine.gov.");
        }

        if (message.equalsIgnoreCase("time")) {
            String time = new java.util.Date().toString();
            sendMessage(channel, sender + ": The time is now " + time);
        }

        if (message.toLowerCase().contains("mayor?")) {
            sendMessage(channel, sender + ": Michael F. Brennan ");
        }

	if (message.toLowerCase().contains("governor?")) {
            sendMessage(channel, sender + ": Paul Richard LePage ");
        }

        if (message.toLowerCase().contains("news")) {
            sendMessage(channel, sender + ": Here's some recent local news, http://www.portlandmaine.gov/CivicAlerts.aspx?CID=1,24,16,7,56,9,15,13,34,27,30,18,17,11,19,31,5,58,22,61,40,53,43,44,37,59,42,21,25,35,46,49");
        }

        if (message.toLowerCase().contains("notes")) {
            sendMessage(channel, sender + ": If you would like meeting notes please respond with 'meeting notes'dd/mm/yyyy' ");
        }

        if (message.toLowerCase().contains("bill")) {
            sendMessage(channel, sender + ": You can pay city bills here - https://trx.npspos.com/payapp/public/ECSale.html?siteId=18545&deptId=18545&urlKey=878a55b229b58de23645e5cfa8ec4f6a60469a37 ");
        }

        if (message.toLowerCase().contains("building code")) {
            sendMessage(channel, sender + ": Learn more about Portlands Building codes here - http://www.portlandmaine.gov/DocumentCenter/View/2325 ");
        }

        if (message.toLowerCase().contains("constitution")) {
            sendMessage(channel, sender + ": You can read the US Constitution here - https://portlandlibrary.bibliocommons.com/item/show/1503083077_the_annotated_us_constitution_and_declaration_of_independence ");
        }

        if (Arrays.asList(badwords).contains(message)) {
            sendMessage(channel, sender + ": Please be civil. ");
        }
    }
}

File: jython/betsyro.py
# Betsyro is a software solution for online civil discourse and civic 
# engagement. It aims to empower those that are unable to be physically present
# at a political venue or otherwise participate in the democratic process.

# Copyright (C) 2015  Scott C. MacCallum
# scm@linux.com

# Copyright (C) 2015  Andrew Scott
# a.clayton.scott@gmail.com

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see http://www.gnu.org/licenses/.

import socket
import optparse

print "[+] Gettings set up"
## set up socket stuff and connect
bot_owner = "Bluebot"
nick = "BlueBot"
chan = "#ppl"
sock = socket.socket()
sock.connect(("irc.freenode.net",6667))
sock.send("USER " + nick + " 0 * :" + bot_owner + "\r\n")
sock.send("NICK " + nick + "\r\n")
print '[+] joined ' + chan
## misc setup


###---------------------------------------------------------
## keep a list of badwords to ding users for
parser = optparse.OptionParser('usage: \n --help <show help> \n -l <bad word list>')
parser.add_option('-l', dest='wl', type='string', help='.txt list of bad words')
(options, args) = parser.parse_args()
wordlist_path = options.wl
wordlist = []

if wordlist_path != None:
	print '[+] bad word list found'
	d = open(wordlist_path)
	for line in d.readlines():
		if (line != "\n" or line != " ") and type(line) == str:
			line = str(line.replace("\n",""))
			wordlist.append(line)
else:
	print '[!] no bad word list set'

## keep track of strikes
user_strikes = {}
def strike(user):
	if user in user_strikes:
		user_strikes[user] += 1
	else:
		user_strikes[user] = 1
	if user_strikes[user] > 2:
		## kick them out
		sock.send("PRIVMSG " + user + " :GET OUT!!!.\r\n")


###--------------------------
## main loop, check the channel for messages, respond to pings, etc.
def main():
	print '[+] starting main loop'
	while 1:
		## keep checking for data
		## reset all the vars after each check
		data = sock.recv(512)
		datalen = len(data.split(" "))
		sender = ""
		msg_type = ""
		msg_rcpt = ""
		message = ""

		###----------------------------------------------
		## reply to keep alive pings
		if data[0:4] == "PING":
			sock.send(data.replace("PING", "PONG"))
		if data[0]!=':':
			continue
		if data.split(" ")[1] == "001":
			sock.send("MODE " + nick + " +B\r\n")
			sock.send("JOIN " + chan + "\r\n")


		###-----------------------------------------------
		##split and assign data parts

		## parse out the actual sender
		send_data = data.split(" ")[0]
		send_data = send_data[1:]
		sender = send_data.split('!')[0]

		## mode
		msg_type = data.split(" ")[1]

		## if it exists get the recipient (room or private)
		if datalen > 2:
			msg_rcpt = data.split(" ")[2]

		## get the actual message body	
		if datalen > 3:
			message = (" ".join(data.split(" ")[3:])).lower()[1:]


		###----------------------------------------------

		## Start replying to private messages	
		if msg_type == "PRIVMSG" and msg_rcpt == nick:
			print "message received"
			print message
			if "mayor" in message:
				sock.send("PRIVMSG " + sender + " :Portland's mayor is Michael F. Brennan\r\n")
			elif "governor" in message:
				sock.send("PRIVMSG " + sender + " :Maine's governor is Paul LePage.. Ew\r\n")
			else:
				sock.send("PRIVMSG " + sender + " :Sorry, I don't understand you...yet.\r\n")

		if msg_type == "PRIVMSG" and msg_rcpt == chan:
			print message
			for curse in wordlist:
				if curse in message:
					print "bad word"
					strike(msg_rcpt)
					sock.send("PRIVMSG " + chan + " :" + sender+ " Please be Civil!.\r\n")
					break


if __name__=='__main__':
	main()




	
